### PR TITLE
Implements kubectl cloudflow deploy blueprint based on verify implementation

### PR DESCRIPTION
**Pull request by debasishg**
### What changes were proposed in this pull request?

* Implements `kubectl-cloudflow deploy --blueprint` 

### Why are the changes needed?

* To support CLI based deploy of blueprint with multi-images

### Does this PR introduce any user-facing change?

* Yes, CLI based deploy of blueprint

### How was this patch tested?

* Tested manually on a GKE cluster

### CODE REVIEW

File: kubectl-cloudflow/cmd/deploy.go [hunk 1]
```diff
@@ -81,22 +105,156 @@ the stored credentials.
...
+func deployBlueprint(cmd *cobra.Command, opts *deployOptions, args []string) {
+	// read blueprint file and load its content
+	blueprintFile := cmd.Flag("blueprint").Value.String()
+	contents, err := fileutils.GetFileContents(blueprintFile)
+	contents = strings.TrimSpace(contents)
+	if err != nil {
+		util.LogAndExit("Failed to fetch blueprint contents from %s", blueprintFile)
+	}
+
+	blueprint, err := verify.VerifyBlueprint(contents)
```
Comment (skonto) [inline kubectl-cloudflow/cmd/deploy.go]: +1

File: kubectl-cloudflow/cloudflowapplication/cloudflowapplication.go [hunk 2]
```diff
@@ -46,7 +46,7 @@ type Endpoint struct {
 // Deployment contains a streamlet deployment
 type Deployment struct {
 	ClassName     string                  `json:"class_name"`
-	Config        json.RawMessage         `json:"config"`
+	Config        json.RawMessage         `json:"config,omitempty"`
```
Comment (RayRoestenburg) [inline kubectl-cloudflow/cloudflowapplication/cloudflowapplication.go]: Did you test this? I just remember that we did an emty config `{}` instead, I'm fine with it, just vaguely remember an issue with this, maybe on the scala side..

Comment (RayRoestenburg) [reply kubectl-cloudflow/cloudflowapplication/cloudflowapplication.go]: It's also possible that we always fill it in on the golang side, but then it is better not to add omitempty here. (also not sure why this change is needed, it changes the contract, but I think not necessary in this change?)

File: kubectl-cloudflow/cloudflowapplication/cloudflowapplication.go [hunk 3]
```diff
@@ -142,6 +141,12 @@ type CloudflowApplicationSpec struct {
 	LibraryVersion string            `json:"library_version,omitempty"`
 }
 
+// Version of the application descriptor format
+const Version = "1"
```
Comment (RayRoestenburg) [inline kubectl-cloudflow/cloudflowapplication/cloudflowapplication.go]: Maybe add these versions (Version and LibraryVersion) in version.go? (Version as ApplicationDescriptorVersion) a lot of other version stuff is there, probably easy to find replace in build.

File: kubectl-cloudflow/cmd/deploy.go [hunk 4]
```diff
@@ -42,9 +46,28 @@ func init() {
 		Use:   "deploy",
 		Short: "Deploys a Cloudflow application to the cluster.",
 		Long: `Deploys a Cloudflow application to the cluster.
-The arguments to the command consists of a docker image path and optionally one
-or more '[streamlet-name].[configuration-parameter]=[value]' pairs, separated by
-a space.
+The behavior of the command depends upon the flags passed in to it. The command supports deployment of
```
Comment (RayRoestenburg) [inline kubectl-cloudflow/cmd/deploy.go]: great stuff 👍

File: kubectl-cloudflow/cmd/deploy.go [hunk 5]
```diff
@@ -81,22 +105,156 @@ the stored credentials.
...
+	}
+}
+
+func deployBlueprint(cmd *cobra.Command, opts *deployOptions, args []string) {
+	// read blueprint file and load its content
+	blueprintFile := cmd.Flag("blueprint").Value.String()
+	contents, err := fileutils.GetFileContents(blueprintFile)
+	contents = strings.TrimSpace(contents)
+	if err != nil {
+		util.LogAndExit("Failed to fetch blueprint contents from %s", blueprintFile)
```
Comment (RayRoestenburg) [inline kubectl-cloudflow/cmd/deploy.go]:
```suggestion
		util.LogAndExit("Failed to read blueprint from %s", blueprintFile)
```

File: kubectl-cloudflow/cmd/deploy.go [hunk 6]
```diff
@@ -81,22 +105,156 @@ the stored credentials.
...
+	} else {
+		deployBlueprint(cmd, opts, args)
+	}
+}
+
+func deployBlueprint(cmd *cobra.Command, opts *deployOptions, args []string) {
+	// read blueprint file and load its content
+	blueprintFile := cmd.Flag("blueprint").Value.String()
+	contents, err := fileutils.GetFileContents(blueprintFile)
+	contents = strings.TrimSpace(contents)
```
Comment (RayRoestenburg) [inline kubectl-cloudflow/cmd/deploy.go]: minor: Why is this necessary?

File: kubectl-cloudflow/cmd/deploy.go (hunk 1 above)
Comment (RayRoestenburg) [reply kubectl-cloudflow/cmd/deploy.go]: At this point you must have already pulled all images, read all labels. It seems that CreateApplicationSpecFromBlueprintAndImages pulls images again and gets labels again? This can probably be improved.

File: kubectl-cloudflow/cmd/deploy.go [hunk 7]
```diff
@@ -81,22 +105,156 @@ the stored credentials.
...
+	}
+
+	blueprint, err := verify.VerifyBlueprint(contents)
+	if err != nil {
+		util.LogAndExit("Blueprint verification failed. Error: %s", err.Error())
+	}
+
+	util.PrintSuccess("Blueprint verified for application %s .. Proceeding with deployment\n", blueprint.GetName())
+
+	applicationSpec, pulledImages, err := deploy.CreateApplicationSpecFromBlueprintAndImages(blueprint, opts.replicasByStreamletName)
```
Comment (RayRoestenburg) [inline kubectl-cloudflow/cmd/deploy.go]: This should take the blueprint and the streamlet descriptors that have already been pulled and verified.

File: kubectl-cloudflow/cmd/deploy.go [hunk 8]
```diff
@@ -81,22 +105,156 @@ the stored credentials.
...
+	if err != nil {
+		util.LogAndExit("Failed to create cloudflow application spec from blueprint and images: %s", err.Error())
```
Comment (RayRoestenburg) [inline kubectl-cloudflow/cmd/deploy.go]: We should always provide feedback in terms the user understands. The user does not know the "application spec". I also expect that the creation of the application spec, once blueprint is verified, should really not fail? what is a good reason?

File: kubectl-cloudflow/cmd/deploy.go [hunk 9]
```diff
@@ -81,22 +105,156 @@ the stored credentials.
...
+	}
+
+	namespace := applicationSpec.AppID
+	k8sClient, k8sErr := k8s.GetClient()
```
Comment (RayRoestenburg) [inline kubectl-cloudflow/cmd/deploy.go]: Might want to move the connect to k8s parts all the way up (getting the clients), especially if the command currently  takes very long (better to fail fast on not getting to the cluster first, then do long process)

File: kubectl-cloudflow/cmd/deploy.go [hunk 10]
```diff
@@ -81,22 +105,156 @@ the stored credentials.
...
+	if k8sErr != nil {
+		util.LogAndExit("Failed to connect to kubernetes cluster %s", k8sErr.Error())
+	}
+
+	cloudflowApplicationClient, err := k8s.GetCloudflowApplicationClient(namespace)
+	if err != nil {
+		util.LogAndExit("Failed to create new kubernetes client with cloudflow config `%s`, %s", namespace, err.Error())
```
Comment (RayRoestenburg) [inline kubectl-cloudflow/cmd/deploy.go]:
```suggestion
		util.LogAndExit("Failed to connect to kubernetes cluster %s", k8sErr.Error())
```
This is just another k8s client, user does not know difference.

File: kubectl-cloudflow/docker/client.go [hunk 11]
```diff
@@ -128,23 +147,16 @@ func imageMatchesNameAndTag(image types.ImageSummary, imageNameAndTag string) bo
...
+// GetStreamletDescriptorsForImage extracts the configuration of a Cloudflow label from a docker image
 // and the image with digest in a struct
-func GetCloudflowStreamletDescriptorsForImage(cli *client.Client, imageName string) (cloudflowapplication.CloudflowStreamletDescriptorsDigestPair, string) {
-
-	_, pullError := PullImage(cli, imageName)
-	if pullError != nil {
-		util.LogAndExit("Failed to pull image %s: %s", imageName, pullError.Error())
-	}
-
+func GetStreamletDescriptorsForImage(cli *client.Client, imageName string) (cloudflowapplication.CloudflowStreamletDescriptorsDigestPair, string, error) {
```
Comment (skonto) [inline kubectl-cloudflow/docker/client.go]: As I discussed with @agemooij in the past for now we should have this part in:
```
_, pullError := PullImage(cli, imageName)	
	if pullError != nil {	
		util.LogAndExit("Failed to pull image %s: %s", imageName, pullError.Error())	
	}	
```

Comment (debasishg) [reply kubectl-cloudflow/docker/client.go]: fixed @skonto ..

File: kubectl-cloudflow/cmd/deploy.go (hunk 1 above)
Comment (debasishg) [reply kubectl-cloudflow/cmd/deploy.go]: @RayRoestenburg we don't pull images in `CreateApplicationSpecFromBlueprintAndImages` - I use the information from the already filled up `blueprint` which is passed to the function.

File: kubectl-cloudflow/cloudflowapplication/cloudflowapplication.go (hunk 2 above)
Comment (debasishg) [reply kubectl-cloudflow/cloudflowapplication/cloudflowapplication.go]: I tested it and it works - will check again ..

File: kubectl-cloudflow/cloudflowapplication/cloudflowapplication.go (hunk 3 above)
Comment (debasishg) [reply kubectl-cloudflow/cloudflowapplication/cloudflowapplication.go]: +1 ..

File: kubectl-cloudflow/cmd/deploy.go (hunk 7 above)
Comment (debasishg) [reply kubectl-cloudflow/cmd/deploy.go]: @RayRoestenburg It takes the blueprint which is already filled up - it does not pull the images again

File: kubectl-cloudflow/cmd/deploy.go (hunk 1 above)
Comment (RayRoestenburg) [reply kubectl-cloudflow/cmd/deploy.go]: It uses getStreamletDescriptorsAndImageInformation, which does the image list again, which is probably already done before?
